File: bst.c
//implement binary search tree
#include<stdio.h>
#include<stdlib.h>
struct node {
	struct node* leftchild;
	struct node* rightchild;
	int data;
	};

struct node*  Insert(struct node* root, int value){
	struct node *newnode = (struct node*)malloc(sizeof(struct node));
	newnode->data=value;
	newnode->leftchild=NULL;
	newnode->rightchild=NULL;
	
	if (root=NULL){
		root=newnode;
		return root;
	}
	else{
		if (value<root->data)
			newnode->leftchild=Insert(newnode->leftchild,value);
		else
			newnode->rightchild=Insert(newnode->rightchild,value);
	}
	return root;
}

		
struct node* Search(struct node* root,int value){
	struct node *curent=root;
	while (current->data!=NULL){
		if (current->data > value)
			current=current->leftchild;
		else if(current->data<value)
			current=current->rightchild;
		if (current==NULL)
			printf("number not found");
	}
	return current;
}
int main(){
	struct node* root=NULL;
	Insert(root,4);
	Insert(root,5);

}

		

File: bst.java
import java.util.*;
import java.io.*;
class node
{
	int data;
	node left;
	node right;
	public node(int data)
	{
		this.data=data;
		left=null;
		right=null;

	}
}
public class bst{
	public static node root;
	public bst(){
		this.root=null;
	}
	public boolean find(int data){
		node current=root;
		while (current!=null)
		{
			if (current.data=data)
			{
				return true;
			}
			else if(current.data<data)
			{
				current=current.right;
			}
			else if(current.data>data)
			{
				current=current.left;
			}
			
		}
		return false;
	}
	public insert(int data)
	{
		node newnode=new node(data);
		if (root==null)
		{
			root=newnode;
		}
		node parent=null;
		node current=newnode;
		while (true)
		{
			parent=current;
			if (current.data<data)
			{
				current=current.right;
				if (current==null)
				{
					parent.right=newnode;
				}
			}
			else
			{
				current=current.left;
				if (current=null)
				{
					parent.left=newnode;
				}
			}
		}
	}

	public boolean delete(int data)
	{
	//find the node
	node parent=root;
	node current=root;
	boolean isLeftChild=false;//find if the current is left child or right
	while (current.data!=data)
	{
		parent=current;
		if (current.data<data)
		{
			isLeftChild=false;
			current=current.right;
		}
		else{
			isLeftChild=true;
			current=current.left;
		}
		if (current=null){return false;}
	
	}
	//we have searched the node and it's parent
	//there will be 3 cases if the node has no child, node has 1 child and node has both 2 childs		
	}

}
